### ISSUE TITLE
Mock large message (>1kb) handling in create blob v2 action

### ISSUE THREAD

**Issue by xernaj** [LogicAppUnit/TestingFramework#24]
Hi,

Has anyone run into an issue with mocking large messages (>1kb) in a create blob v2 action? (macos)

_status='Failed', statusCode='InvalidProtocolResponse', error='{"code":"InvalidProtocolResponse","message":"The response to partial content upload initiating request is not valid. The response to initiating partial content upload request must contain a valid location header."}_

If I mock the "Location" header then this error appears:

{"code":"InvalidProtocolResponse","message":"The response to partial content upload request is not valid. The location header value returned in the response 'https%3A%2F%2Flocalhost%3A7075%2Fchunk' must be a well formed absolute URI not referencing local host or UNC path."}

It appears that chunking can't be mocked so I've started on something that will disable chunking. Hopefully this is the right path?

**Comment by mark-abrams** [comment]
I forgot to add that I'm using a message that is 10Kb in size when saved to blob storage.

**Comment by xernaj** [comment]
Hi Mark, I'll try reproduce that workflow and see what happens on my Mac (Intel).

**Comment by xernaj** [comment]
Hi Mark,

I've created a sample workflow:

- recurrence trigger
- http get a >1kb json file
- http post body from get body


![Screen Shot 2023-07-27 at 9 58 03 pm](https://github.com/LogicAppUnit/TestingFramework/assets/11097415/7b458410-7322-419c-bdfc-3965f8cd8b3f)

I noticed it also appears in managed create blob connector (not in upload built-in connector).
![Screen Shot 2023-07-27 at 9 58 12 pm](https://github.com/LogicAppUnit/TestingFramework/assets/11097415/386f9b4d-19c5-4477-a864-08afe9ba7269)
Notice the "content transfer" flag which appears in JSON as:
```
"runtimeConfiguration": {
    "contentTransfer": {
        "transferMode": "Chunked"
    }
}
```

**Comment by mark-abrams** [comment]
Hi @xernaj , now I understand the issue and I can re-create it myself. Basically you are enabling the [chunking upload protocol](https://learn.microsoft.com/en-us/azure/logic-apps/logic-apps-handle-large-messages#upload-content-in-chunks) in the HTTP action called `POST` and the testing framework is not implementing the "server-side" requirements for the chunking protocol to work.

As a workaround it is possible to implement the server-side requirements in a test case. This example uses a mix of the Fluent API (new in version 1.7.0) and the delegate mock function, since there is some functionality that is needed that cannot be implemented in the Fluent API.

Firstly, set up some variables in the test case to manage state of the chunking process:

![image](https://github.com/LogicAppUnit/TestingFramework/assets/66829458/1cdb2d69-9df2-46d0-9400-4c6e21afec9f)

Then configure a Request Matcher and Response Builder to handle the initial POST request in the chunking protocol:

![image](https://github.com/LogicAppUnit/TestingFramework/assets/66829458/e54315a5-a5b9-4d4a-9566-4b2f2b04088b)

Then configure a delegate mock to handle the PATCH requests that contain the chunks of data:

![image](https://github.com/LogicAppUnit/TestingFramework/assets/66829458/81b72942-fb7c-44de-9063-4b7b62d5eb0e)

Once this is done, you can easily assert the contains of the stream, i.e. the data that was sent by the `POST` action in the workflow:

![image](https://github.com/LogicAppUnit/TestingFramework/assets/66829458/121ddae7-c034-4e01-aa23-918d3988e917)

I realise that this is a bit fiddly to set up, but at least this is a workaround that works.

**Comment by mark-abrams** [comment]
There are a few ways that this could be improved:

1. Change the framework so that the `runtimeConfiguration.contentTransfer` setting for all HTTP actions are removed from the workflow definition before the test is run. This appraoch would be similar to how the retry policies for HTTP actions are updated to `none`. Might want to enable/disable this feature via a test configuration setting.
2. Change the Fluent API processing so that the "server-side" requirements for the chunking protocol are implemented internally and not visible to the test author. Maybe this additional logic could be enabled using a new `EnableChunking()` method in the fluent API.

Option 1 is the easiest to implement, but I'm very careful about making changes to the workflow definition. Every change to the definition moves the workflow being tested away from the workflow that would run in a production environment. This is why option 2 might be better - there would be no change required to the workflow definition and the testing framework would natively handle the chunking protocol, in a similar way to the workaround I have shown.

Also need to think if there are any test scenarios where the enabling of chunking for a HTTP action would be required. I can't think of any currently, since the chunking in the workflow is "out of the box" Logic App functionality and therefore (probably) would not be part of any test scenario.

**Comment by xernaj** [comment]
Thanks Mark, I'll give that a go. Last time I tried to mock the chunking prior to v1.70, the function app runtime complained about it not allowing chunking when calling localhost or a non FQDN - hopefully this works.

> Hi @xernaj , now I understand the issue and I can re-create it myself. Basically you are enabling the [chunking upload protocol](https://learn.microsoft.com/en-us/azure/logic-apps/logic-apps-handle-large-messages#upload-content-in-chunks) in the HTTP action called `POST` and the testing framework is not implementing the "server-side" requirements for the chunking protocol to work.
> 
> As a workaround it is possible to implement the server-side requirements in a test case. This example uses a mix of the Fluent API (new in version 1.7.0) and the delegate mock function, since there is some functionality that is needed that cannot be implemented in the Fluent API.
> 
> I realise that this is a bit fiddly to set up, but at least this is a workaround that works.

**Comment by xernaj** [comment]
Hi @mark-abrams ,

Thanks for your effort with v1.70 fluent api - (it looks nice and clean) and looking at the chunking issue. I couldn't get your workaround to succeed on my intel mac, with func core tools "4.0.5198". On macos, it seems that the function app runtime doesn't support localhost in the `Location` header.

_"{"code":"InvalidProtocolResponse","message":"The response to partial content upload request is not valid. The location header value returned in the response 'http://localhost:7075/api/v1.1/287124d3-a604-4b35-97fa-2cc2d7332772' must be a well formed absolute URI not referencing local host or UNC path."}'"_

Here's the repo and sample test https://github.com/xernaj/TestingFramework/blob/b721c903ba5ce864e5817e31eb9d689576813f23/LogicAppUnit.Samples.LogicApps.Tests/ChunkedTransferWorkflowTest/ChunkedTransferWorkflowTest.cs#L40

I think I will look at implementing your Option 1 (user switchable disabling of chunking by modifying workflow.json) when/if time allows. My only workaround is to disable chunking - it appears to be auto enabled for HTTP and managed connector blob actions so there's a lot of workflows I may need to update to make them unit testable.

Cheers

**Comment by mark-abrams** [comment]
Hi @xernaj , I can re-produce your problem on both Ubuntu and MacOS. It seems that the URL in the `Location` header cannot reference `localhost`, or even `127.0.01`. It needs to use a DNS name, even if this is the name of the host host server. I tried changing the code I shared with you earlier to use `Environment.MachineName` and this didn't work either - same error.

![image](https://github.com/LogicAppUnit/TestingFramework/assets/66829458/db6e6134-6146-4948-92cf-cd836c9ac658)

But if I change the core LogicAppUnit code so that the mock HTTP Server listens on a URL created using `Environment.MachineName` and not `localhost` it works for Windows and Ubuntu, but not for macOS! Take a look at [this action workflow build](https://github.com/LogicAppUnit/TestingFramework/actions/runs/5962204165) where the tests are successful for Windows and Ubuntu, but failed on macOS with this error:

```
System.AggregateException: One or more errors occurred. (nodename nor servname provided, or not known (mac-1692870849291:7071)) ---> System.Net.Http.HttpRequestException: nodename nor servname provided, or not known (mac-1692870849291:7071) ---> System.Net.Sockets.SocketException: nodename nor servname provided, or not known
```

I'm not sure why this is failing on macOS. I was hoping to add support to the testing framework and the fluent API so that it automatically supports chunking in the mock HTTP server, but I'm not sure how to get this working with macOS if neither `localhost` nor `Environment.MachineName` work in the URL for the `Location` header. The only other option is to pre-process the workflow definition file and remove all of the chunking options for HTTP actions, but I was hoping to avoid this. If I can't think of a better solution, I will implement this in the next version (1.8.0).
